src/operator/contrib/rroi_align.cc

```diff
@@ -0,0 +1,316 @@
+struct position_for_bilinear_interpolate {
+  // 4 positions and corresponding weights for
+  // computing bilinear interpolation
+  int pos1, pos2, pos3, pos4;
+  DType w1, w2, w3, w4;
+};
+
+template <typename DType>
+void pre_calc_for_bilinear_interpolate(
+    const int height, const int width, const int pooled_height, const int pooled_width,
+    const int iy_upper, const int ix_upper, DType roi_start_h, DType roi_start_w,
+    DType bin_size_h, DType bin_size_w, int roi_bin_grid_h, int roi_bin_grid_w,
+    DType roi_center_h, DType roi_center_w, DType theta,
+    std::vector<position_for_bilinear_interpolate<DType>> *pre_calc) {
+  int pre_calc_index = 0;
+  DType cosTheta = cos(theta);
+  DType sinTheta = sin(theta);
+  for (int ph = 0; ph < pooled_height; ph++) {
+    for (int pw = 0; pw < pooled_width; pw++) {
+      // calc bin grid position (xx,yy)
+      for (int iy = 0; iy < iy_upper; iy++) {
+        const DType yy = roi_start_h + ph * bin_size_h +
+            static_cast<DType>(iy + .5f) * bin_size_h /
+                static_cast<DType>(roi_bin_grid_h);  // e.g., 0.5, 1.5
+        for (int ix = 0; ix < ix_upper; ix++) {
+          const DType xx = roi_start_w + pw * bin_size_w +
+              static_cast<DType>(ix + .5f) * bin_size_w /
+                  static_cast<DType>(roi_bin_grid_w);
+
+          // Rotate by theta around the center and translate
+          DType x = xx * cosTheta + yy * sinTheta + roi_center_w;
+          DType y = yy * cosTheta - xx * sinTheta + roi_center_h;
+
+          // deal with: inverse elements are out of feature map boundary
+          if (y < -1.0 || y > height || x < -1.0 || x > width) {
+            // empty
+            position_for_bilinear_interpolate<DType> pc;
+            pc.pos1 = 0;
+            pc.pos2 = 0;
+            pc.pos3 = 0;
+            pc.pos4 = 0;
+            pc.w1 = 0;
+            pc.w2 = 0;
+            pc.w3 = 0;
+            pc.w4 = 0;
+            pre_calc->at(pre_calc_index) = pc;
+            pre_calc_index += 1;
+            continue;
+          }
+          if (y <= 0) {
+            y = 0;
+          }
+          if (x <= 0) {
+            x = 0;
+          }
+
+          // calc 4 points for interpolation
+          int y_low = static_cast<int>(y);
+          int x_low = static_cast<int>(x);
+          int y_high;
+          int x_high;
+          if (y_low >= height - 1) {
+            y_high = y_low = height - 1;
+            y = (DType)y_low;
+          } else {
+            y_high = y_low + 1;
+          }
+          if (x_low >= width - 1) {
+            x_high = x_low = width - 1;
+            x = (DType)x_low;
+          } else {
+            x_high = x_low + 1;
+          }
+          DType ly = y - y_low;
+          DType lx = x - x_low;
+          DType hy = 1. - ly, hx = 1. - lx;
+          DType w1 = hy * hx, w2 = hy * lx, w3 = ly * hx, w4 = ly * lx;
+
+          // Save weights and indices
+          position_for_bilinear_interpolate<DType> pc;
```

Comment:
`position_for_bilinear_interpolate<DType> &pc = (*pre_calc)[pre_calc_index];`
There will be no extra copy, and access by index is faster than .at()

Reply:
thank you for your advices, @wkcn.